src/components/Microsoft.DotNet.UpgradeAssistant.Analysis/AnalyzeResultWriter.cs

```diff
@@ -21,6 +21,23 @@ public AnalyzeResultWriter(ISerializer serializer)
             this._serializer = serializer;
         }
 
+        public async Task WriteAsync(IAsyncEnumerable<AnalyzeResultDefinition> results, CancellationToken token)
+        {
+            if (results is null)
+            {
+                throw new ArgumentNullException(nameof(results));
+            }
+
+            var sarifLog = new SarifLog()
+            {
+                Version = SarifVersion.Current,
+                SchemaUri = new Uri("http://json.schemastore.org/sarif-2.1.0"),
+                Runs = await ExtractRunsAsync(results).ConfigureAwait(false),
+            };
+
+            _serializer.Write(_sarifLogPath, sarifLog);
+        }
+
         private static async Task<List<Run>> ExtractRunsAsync(IAsyncEnumerable<AnalyzeResultDefinition> analyzeResults)
```

Comment:
IAsyncEnumerable would make more sense here and you can just `yield return` the result